### PR TITLE
Options and Keywords clean-up

### LINKED ISSUE #235: No method for `optimize(f, x, OptimizationOptions(...))`
No method for `optimize(f, x, OptimizationOptions(...))`,  `optimize(f, g!, x, OptimizationOptions(...))` or for `optimize(f, g!, h!, x, OptimizationOptions(...))`. These should just default to Nelder-Mead, BFGS, and Newton as usual.

### PR DESCRIPTION
**Pull request by pkofod**
Ref #243 also fixed #235 

- [x] Add tests for fix in #235 
- [x] See what do to deprecation-wise
- [x] Deprecate `linesearch!` and `precondprep!` in favor of `linesearch` and `precondprep`
- [x] <s>Use "manual" deprecation to deprecate the keyword methods so there is no deprecation warning given when keywords are not used.</s> parse and pass keywords in a better way

**Summary**:
This PR removes ! in keyword argument names to avoid the situation where a user write `kw!=kw_value` and gets an error. In this PR the user would write `kw=kw_value` and be fine. We don't want an interface that depends on white space, when white space is generally not relied upon in Julia.

Additionally, this PR seeks to remove the exported `OptimizationOptions` in favor of the unexported `(Optim.)Options`. This change also seeks to remove all keyword arguments that used to be passed to `optimize(args...;kwargs...)`, such that it is required for a user to construct the `Options` and pass it to `optimize` as a positional argument. This seems more in line with the general use of dispatch in Julia, and allows us to parameterize Options later if we wish, without having to maintain two different APIs. This interface might seem annoying at first (you *have* to write `Optim.Options(` to set an option), but since users had to write `method=...` before to be able to use the other keywords, the overall keystroke count is basically the same.

### CODE REVIEW

**Comment by pkofod** [comment]
Does anyone object to this change? I don't want to spend time on the deprecations if people are against it anyway.

**Comment by anriseth** [comment]
I think the proposed change is good

**Comment by pkofod** [comment]
Took care of #285  while cleaning up options in the latest commit.

File: docs/src/dev/contributing.md [hunk 1]
```diff
@@ -18,12 +18,12 @@ that does the actual work. Say you want to contribute a solver called
 
 ```
 immutable Minim{T} <: Optimizer
-    linesearch!::Function
+    linesearch::Function
```
Comment (anriseth) [inline docs/src/dev/contributing.md]: I think we should still have the `linesearch!` in the type fields. The issue we want to fix is the problem that arises in the keywords.

File: src/accelerated_gradient_descent.jl [hunk 2]
```diff
@@ -8,11 +8,11 @@
 
 
 immutable AcceleratedGradientDescent <: Optimizer
-    linesearch!::Function
+    linesearch::Function
```
Comment (anriseth) [inline src/accelerated_gradient_descent.jl]: Keep it `linesearch!` ?

File: src/accelerated_gradient_descent.jl [hunk 3]
```diff
@@ -61,7 +61,7 @@ function update_state!{T}(d, state::AcceleratedGradientDescentState{T}, method::
 
     # Determine the distance of movement along the search line
     state.alpha, f_update, g_update =
-      method.linesearch!(d, state.x, state.s, state.x_ls, state.g_ls, state.lsr, state.alpha, state.mayterminate)
+      method.linesearch(d, state.x, state.s, state.x_ls, state.g_ls, state.lsr, state.alpha, state.mayterminate)
```
Comment (anriseth) [inline src/accelerated_gradient_descent.jl]: Keep it `linesearch!`? This is a method that should tell the person reading the code that some of the objects in its argument will be changed.

File: src/deprecate.jl [hunk 4]
```diff
@@ -18,8 +18,8 @@ for name in names(LineSearches)
     eval(:(@deprecate $name LineSearches.$name))
 end
 
-@deprecate hz_linesearch! LineSearches.hagerzhang!
-@deprecate mt_linesearch! LineSearches.morethuente!
-@deprecate interpolating_linesearch! LineSearches.strongwolfe!
-@deprecate backtracking_linesearch! LineSearches.backtracking!
-@deprecate interpbacktracking_linesearch! LineSearches.interpbacktracking!
+@deprecate hz_linesearch LineSearches.hagerzhang!
```
Comment (pkofod) [inline src/deprecate.jl]: this is a mistake...

**Comment by pkofod** [comment]
> I think we should keep the type objects in the optimizers as linesearch! still. The issue was to have the ! in kwargs

Yes, it is indeed a problem to have the ! in kwargs only, but isn't it odd to have a keyword for a field being different than the actual field name? I guess it's okay if spelled out somewhere, it would probably just surprise most people if they looked at the type definition, saw a field, and then couldn't set that field with the keyword spelled the same.

**Comment by timholy** [comment]
FYI, going forward it might be better to declare these types like this:
```jl
immutable AcceleratedGradientDescent{LS<:Function} <: Optimizer
    linesearch!::LS
end
```
The reason is that in julia 0.5 and above, each function has its own type. Using the type parameter means this type is fully type-stable, and you can even dispatch like this:
```jl
foo(o::Optimizer{typeof(hagerzhang!)}) = 1
foo(o::Optimizer) = 2
```

**Comment by anriseth** [comment]
> Yes, it is indeed a problem to have the ! in kwargs only, but isn't it odd to have a keyword for a field being different than the actual field name? I guess it's okay if spelled out somewhere, it would probably just surprise most people if they looked at the type definition, saw a field, and then couldn't set that field with the keyword spelled the same.

Both options complicate things I guess. My perspective is the following: To understand what a solver does, I will look at the code in the `optimize` method. And if I see `method.linesearch!`, I understand that the call to line search may change my arguments. If it only says `method.linesearch`, I would expect it not to. But the line searches will change the LineSearchResults object, and potentially df (if they include evaluation counts).

**Comment by pkofod** [comment]
> FYI, going forward it might be better to declare these types like this:
> 
> immutable AcceleratedGradientDescent{LS<:Function} <: Optimizer
>     linesearch!::LS
> end
> 
> The reason is that in julia 0.5 and above, each function has its own type. Using the type parameter means this type is fully type-stable, and you can even dispatch like this:
> 
> foo(o::Optimizer{typeof(hagerzhang!)}) = 1
> foo(o::Optimizer) = 2

Thank you for the comment. I just realized that fact yesterday, I only thought it applied to anonymous functions for some reason. I should probably revisit some of the changes made to nelder mead over the summer, where I think I might have over-complicated some things for this very reason.

**Comment by pkofod** [comment]
> Both options complicate things I guess. My perspective is the following: To understand what a solver does, I will look at the code in the optimize method. And if I see method.linesearch!, I understand that the call to line search may change my arguments. If it only says method.linesearch, I would expect it not to. But the line searches will change the LineSearchResults object, and potentially df (if they include evaluation counts).

I'll add the ! to the field names, and only change the keywords.

**Comment by pkofod** [comment]
precondprep should suffer the same fate (note to self)

**Comment by KristofferC** [comment]
Might be worth thinking if deprecated keywords should just be warned about once per session.

**Comment by pkofod** [comment]
> Might be worth thinking if deprecated keywords should just be warned about once per session.

True. On the other hand, I would prefer if people fixed it right away ;)

**Comment by pkofod** [comment]
@timholy @johnmyleswhite is this change OK with you? I'm thinking about the `OptimizationOptions -> Optim.Options` and removal of the keyword args.

**Comment by pkofod** [comment]
Removed the "controversial" keywords deprecation, so I'm just going to merge this. Keywords can be deprecated by changing two lines, so we can still do it by next tag if someone feels like we should. The new "low maintenance" way of doing it is fine by me at least.

**Pull request merged**